

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>RNA 3D Structure Using forgi.threedee &#8212; forgi 2.0.0 documentation</title>
    <link rel="stylesheet" href="_static/haiku.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/print.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/theme_extras.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Forgi API documentation" href="apidoc/forgi.html" />
    <link rel="prev" title="RNA Secondary Structure as a Graph Using the forgi Library" href="graph_tutorial.html" /> 
  </head>
  <body>
      <div class="header"><h1 class="heading"><a href="index.html">
          <span>forgi 2.0.0 documentation</span></a></h1>
      </div>
      <div class="topnav">
      
        <p>
        «&#160;&#160;<a href="graph_tutorial.html">RNA Secondary Structure as a Graph Using the forgi Library</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="apidoc/forgi.html">Forgi API documentation</a>&#160;&#160;»
        </p>

      </div>
      <div class="content">
        
        
  <div class="section" id="rna-3d-structure-using-forgi-threedee">
<span id="forgi-threedee-tutorial"></span><h1>RNA 3D Structure Using forgi.threedee<a class="headerlink" href="#rna-3d-structure-using-forgi-threedee" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">forgi.threedee</span></code> is an extension of <code class="docutils literal"><span class="pre">forgi</span></code> capable of handling 3D data
about RNA structures. It provides methods for extracting secondary structure as
well as creating a coarse grain representation of 3D RNA structures.</p>
</div>
<div class="section" id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h2>
<p>We strongly recommend installing MC_Annotate or DSSR when using forgi to work
with PDB files. See <a class="reference internal" href="download.html#ext-dep"><span class="std std-ref">Optional, external dependencies (require manual installation)</span></a></p>
</div>
<div class="section" id="extracting-2d-structure-from-a-3d-structure">
<h2>Extracting 2D structure from a 3D structure<a class="headerlink" href="#extracting-2d-structure-from-a-3d-structure" title="Permalink to this headline">¶</a></h2>
<a class="reference internal image-reference" href="_images/3d_to_2d.png"><img alt="_images/3d_to_2d.png" class="align-center" src="_images/3d_to_2d.png" style="width: 500px;" /></a>
<p>To extract the base-pair information from a 3D structure stored in a PDB or
MMCIF file, just convert the file to a fasta file using the <code class="code docutils literal"><span class="pre">rnaConvert.py</span></code>
script:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ rnaConvert.py -T fasta examples/1y26.pdb
&gt;1y26
CGCUUCAUAUAAUCCUAAUGAUAUGGUUUGGGAGUUUCUACCAAGAGCCUUAAACUCUUGAUUAUGAAGUG
(((((((((...(((((((.......)))))))........((((((.......))))))..)))))))))
</pre></div>
</div>
<p>This can also be done programmatically in python:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cg</span><span class="p">,</span> <span class="o">=</span> <span class="n">forgi</span><span class="o">.</span><span class="n">load_rna</span><span class="p">(</span><span class="s1">&#39;test/forgi/threedee/data/1y26.pdb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span> <span class="n">cg</span><span class="o">.</span><span class="n">to_fasta_string</span><span class="p">()</span> <span class="p">)</span>
<span class="go">&gt;1y26_X</span>
<span class="go">CGCUUCAUAUAAUCCUAAUGAUAUGGUUUGGGAGUUUCUACCAAGAGCCUUAAACUCUUGAUUAUGAAGUG</span>
<span class="go">((((((((((..((((((.........))))))......).((((((.......))))))..)))))))))</span>
</pre></div>
</div>
<p>Pseudoknots are removed using the knotted2nested.py script which is included
with the generous permition of Dr. Sandra Smit <a class="reference internal" href="#ref1"><span class="std std-ref">[1]</span></a>. Users making use of this
feature of the 3D-to-2D facilities of the <code class="docutils literal"><span class="pre">forgi.threedee</span></code> package should
cite the articles listed in the citations section at the bottom of this page.</p>
<p>To keep pseudoknots, use the <code class="code docutils literal"><span class="pre">--pseudoknots</span></code> commandline option to rnaConvert.py
or <code class="code docutils literal"><span class="pre">pdb_remove_pk=False</span></code> in the <code class="code docutils literal"><span class="pre">load_rna</span></code> function.</p>
</div>
<div class="section" id="creating-a-coarse-grain-3d-representation-of-an-rna-molecule">
<h2>Creating a Coarse Grain 3D Representation of an RNA Molecule<a class="headerlink" href="#creating-a-coarse-grain-3d-representation-of-an-rna-molecule" title="Permalink to this headline">¶</a></h2>
<p>One can imagine an RNA molecule as a collection of helices and … not helices
(hairpins, interior loops, multiloops, etc..) as described in the
<a class="reference internal" href="graph_tutorial.html#forgi-graph-tutorial"><span class="std std-ref">RNA Secondary Structure as a Graph Using the forgi Library</span></a>. By creating an abstraction for the canonical
helixes and representing them as cylinders, we can create a coarse grain
representaiton of the RNA molecule:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ python rnaConvert.py -T forgi 2l94.pdb
name 2l94_A
length 45
seq GGGAAGAUCUGGCCUUCCCACAAGGGAAGGCCAGGGAAUCUUCCC
seq_ids A:1 A:2 A:3 A:4 A:5 A:6 A:7 A:8 A:9 A:10 A:11 A:12 A:13 A:14 A:15 A:16 A:17 A:18 A:19 A:20 A:21 A:22 A:23 A:24 A:25 A:26 A:27 A:28 A:29 A:30 A:31 A:32 A:33 A:34 A:35 A:36 A:37 A:38 A:39 A:40 A:41 A:42 A:43 A:44 A:45
define s0 1 8 38 45
define i0 35 37
define s1 9 19 24 34
define h0 20 23
connect s0 i0
connect s1 i0 h0
coord s1 -3.1866813709154620 3.4914107521504016 3.6351776728651553 21.4829112410527117 -2.1083990492427773 0.5057319475413644
coord s0 -25.6897073719667972 -5.8096386325483751 -4.0265189300856425 -10.6552743590351255 3.1029175167846290 -0.4045073529842251
coord h0 21.4829112410527117 -2.1083990492427773 0.5057319475413644 29.0380001068115234 4.4409999847412109 -5.3439998626708984
twist s1 -0.0184282079893236 0.4246648355404593 -0.9051630674114455 -0.1571049415641848 -0.1462958719913227 -0.9766860064393329
twist s0 -0.5333577254411702 0.8264653710187239 0.1802346448913319 0.4343359029793392 -0.4058832531020076 -0.8041213268123496
interacting A:28
interacting A:17
interacting A:11
interacting A:16
interacting A:10
interacting A:15
interacting A:9
interacting A:14
interacting A:34
interacting A:12
interacting A:33
interacting A:27
interacting A:32
interacting A:26
interacting A:31
interacting A:25
interacting A:30
interacting A:29
vres i0 -2.31427602 1.57444474 3.54455842 -7.71669904 4.28261433 4.39684139 -5.20558252 8.97808845 5.88005184
vres h0 -2.11144177 4.89881497 -3.76862411 0.26981184 0.20644734 -4.70205156 5.79209284 -0.00000000 0.00000000 3.79477715 2.83408417 3.24575331
</pre></div>
</div>
<p>The file format is an extention of the BulgeGraph file format
(<a class="reference internal" href="graph_tutorial.html#forgibgformat"><span class="std std-ref">The forgi fileformat for secondary structure</span></a>).
The lines beginning with <code class="code docutils literal"><span class="pre">coord</span></code> indicate the positions of the helices. They
are calculated by fitting a cylinder to each helical region of the 3D
structure.</p>
<p>The <code class="code docutils literal"><span class="pre">seq_ids</span></code> line is used to store the residue numbers used in the PDB
prefixed by the chain (For more on residue numbering see …).
The <code class="code docutils literal"><span class="pre">interacting</span></code> lines give residues that interact with a non-RNA molecule,
in this case a ligand. The <code class="code docutils literal"><span class="pre">vres</span></code> lines give the positions of the C1’ atom
of the residues in loops (Always 3 coordinates correspond to one residue).
If the C1’ atom is missing, the coordinates refer to a very approximate
position instead.</p>
</div>
<div class="section" id="visualizing-the-coarse-grain-helical-representation">
<h2>Visualizing the Coarse Grain Helical Representation<a class="headerlink" href="#visualizing-the-coarse-grain-helical-representation" title="Permalink to this headline">¶</a></h2>
<p>The coarse grain representation shown in the prior example can be visualized
using <a class="reference external" href="http://www.pymol.org/">Pymol</a> and the <code class="docutils literal"><span class="pre">visualize_rna.py</span></code> script. The green cylinders below
represent canonical helical regions, red represent multiloops and blue -
hairpin loops:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">visualize_rna</span><span class="o">.</span><span class="n">py</span> <span class="mi">1</span><span class="n">y26</span><span class="o">.</span><span class="n">pdb</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/1y26_visualized.png"><img alt="_images/1y26_visualized.png" class="align-center" src="_images/1y26_visualized.png" style="width: 270px;" /></a>
<p>The <code class="code docutils literal"><span class="pre">visualize_rna.py</span></code> script can also display files in the forgi
file format:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">visualize_rna</span><span class="o">.</span><span class="n">py</span> <span class="n">examples</span><span class="o">/</span><span class="mi">1</span><span class="n">y26</span><span class="o">.</span><span class="n">cg</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/1y26_coarse.png"><img alt="_images/1y26_coarse.png" class="align-center" src="_images/1y26_coarse.png" style="width: 270px;" /></a>
</div>
<div class="section" id="calculate-the-distance-between-two-coarse-grain-elements">
<h2>Calculate the Distance Between Two Coarse-Grain Elements<a class="headerlink" href="#calculate-the-distance-between-two-coarse-grain-elements" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cg</span><span class="p">,</span> <span class="n">forgi</span><span class="o">.</span><span class="n">load_rna</span><span class="p">(</span><span class="s1">&#39;test/forgi/threedee/data/1y26.cg&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dist</span> <span class="o">=</span> <span class="n">cg</span><span class="o">.</span><span class="n">element_physical_distance</span><span class="p">(</span><span class="s1">&#39;h0&#39;</span><span class="p">,</span> <span class="s1">&#39;h1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="p">(</span><span class="n">dist</span><span class="p">)</span>
<span class="go">7.87989954482</span>
</pre></div>
</div>
</div>
<div class="section" id="find-out-how-much-an-interior-loop-bends-a-stem">
<h2>Find out How Much an Interior Loop Bends a Stem<a class="headerlink" href="#find-out-how-much-an-interior-loop-bends-a-stem" title="Permalink to this headline">¶</a></h2>
<p>Interior loops often place kinks in long stem-like structures. This leads
to a change in the base stacking pattern and can indicate functional
relevance. To extract this information for a given PDB file, we need to
iterate over each interior loop and calculate the angle between the two
stems it connects:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">forgi.threedee.utilities.vector</span> <span class="k">as</span> <span class="nn">ftuv</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cg</span> <span class="o">=</span> <span class="n">forgi</span><span class="o">.</span><span class="n">load_rna</span><span class="p">(</span><span class="s1">&#39;test/forgi/threedee/data/1GID_native.pdb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">iloop</span> <span class="ow">in</span> <span class="n">cg</span><span class="o">.</span><span class="n">iloop_iterator</span><span class="p">():</span>
<span class="gp">... </span>    <span class="n">conn</span> <span class="o">=</span> <span class="n">cg</span><span class="o">.</span><span class="n">connections</span><span class="p">(</span><span class="n">iloop</span><span class="p">)</span>
<span class="gp">... </span>    <span class="c1"># conn contains two values [&#39;s0&#39;, &#39;s1&#39;]</span>
<span class="gp">... </span>    <span class="n">angle</span> <span class="o">=</span> <span class="n">ftuv</span><span class="o">.</span><span class="n">vec_angle</span><span class="p">(</span><span class="n">cg</span><span class="o">.</span><span class="n">coords</span><span class="o">.</span><span class="n">get_direction</span><span class="p">(</span><span class="n">conn</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">cg</span><span class="o">.</span><span class="n">coords</span><span class="o">.</span><span class="n">get_direction</span><span class="p">(</span><span class="n">conn</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">iloop</span><span class="p">,</span> <span class="n">angle</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">i3 0.307770762476</span>
<span class="go">i2 2.74681004918</span>
<span class="go">i1 0.1697963999</span>
<span class="go">i0 0.491755788011</span>
<span class="go">i5 0.456253974086</span>
<span class="go">i4 0.261428615896</span>
<span class="go">i7 0.15810445353</span>
<span class="go">i6 0.510919193909</span>
</pre></div>
</div>
<p><code class="code docutils literal"><span class="pre">conn[0]</span></code> and <code class="code docutils literal"><span class="pre">conn[1]</span></code> are the identifiers of the first and second connected stems,
respectively.
<code class="code docutils literal"><span class="pre">cg.coords[conn[0]][0]</span></code> contains the coordinates of the front end
of the first stem and <code class="code docutils literal"><span class="pre">cg.coords[conn[0]][1]</span></code> the end coordinates.
<code class="code docutils literal"><span class="pre">cg.coords.get_direction(conn[0])</span></code> gives the vector from the stem start to the
stem end, i.e.  <code class="code docutils literal"><span class="pre">cg.coords[conn[0]][1]</span> <span class="pre">-</span>&#160; <span class="pre">`cg.coords[conn[0]][0]</span></code>.
We then use these direction vectors to calculate the angle in radians.
This example, using the Group-I intron,
indicates the presence of a kink-turn at interior loop <code class="code docutils literal"><span class="pre">i2</span></code>.</p>
</div>
<div class="section" id="check-if-a-bio-pdb-chain-is-rna">
<h2>Check if a Bio.PDB Chain is RNA<a class="headerlink" href="#check-if-a-bio-pdb-chain-is-rna" title="Permalink to this headline">¶</a></h2>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">Bio</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">forgi.threedee.utilities.pdb</span> <span class="k">as</span> <span class="nn">ftup</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">os.path</span> <span class="k">as</span> <span class="nn">op</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filename</span> <span class="o">=</span> <span class="n">op</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s1">&#39;~/data/pdbs/1Y26.pdb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">structure</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">PDB</span><span class="o">.</span><span class="n">PDBParser</span><span class="p">()</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s1">&#39;blah&#39;</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ftup</span><span class="o">.</span><span class="n">is_rna</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">structure</span><span class="o">.</span><span class="n">get_chains</span><span class="p">())[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</div>
<div class="section" id="calculate-the-rmsd-between-two-pdb-chains">
<h2>Calculate the RMSD Between two PDB Chains<a class="headerlink" href="#calculate-the-rmsd-between-two-pdb-chains" title="Permalink to this headline">¶</a></h2>
<p>The root mean square deviation (RMSD) is a measure of how much two molecules
differ in their atomic coordinates. The value, of course, depends on how the
two molecules are superimposed, but in most cases (including here) a rotation
and translation is applied such that the RMSD is minimized. The RMSD value
is often used to compare the models created by structure prediction software to the real
structure, and can easily be calculated using the <code class="code docutils literal"><span class="pre">pdb_rmsd</span></code> method. It can
take an optional <code class="code docutils literal"><span class="pre">sidechains</span></code> parameter (which defaults to False), to indicate
that the sidechains (bases) should be included in the RMSD calculation. If it
is False, then only the backbone atoms are used in the calculation:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">forgi.threedee.utilities.pdb</span> <span class="k">as</span> <span class="nn">ftup</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">Bio.PDB</span> <span class="k">as</span> <span class="nn">bpdb</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">bpdb</span><span class="o">.</span><span class="n">PDBParser</span><span class="p">()</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s1">&#39;temp&#39;</span><span class="p">,</span> <span class="s1">&#39;test/forgi/threedee/data/2mis.pdb&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">get_chains</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ftup</span><span class="o">.</span><span class="n">pdb_rmsd</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">c</span><span class="p">)</span>
<span class="go">(180, 1.0314194769216807e-14, (array([[  1.00000000e+00,  -1.94289029e-16,   1.11022302e-16],</span>
<span class="go">       [  8.32667268e-17,   1.00000000e+00,   6.93889390e-17],</span>
<span class="go">              [ -5.55111512e-17,   6.93889390e-17,   1.00000000e+00]]), array([ -5.68434189e-14,   2.84217094e-14,  -1.73194792e-14])))</span>
</pre></div>
</div>
<p>The return value is a tuple containing the number of atoms that were
superimposed, the RMSD value and another tuple containing the optimal rotation
matrix and translation vector.</p>
<p>This can also be achieved using the compareRNA.py script:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ compare_RNA.py &#39;test/forgi/threedee/data/2mis.pdb&#39; &#39;test/forgi/threedee/data/2mis.pdb&#39; --pdb-rmsd
PDB-RMSD (chains A):        0.000
</pre></div>
</div>
</div>
<div class="section" id="calculate-the-rmsd-between-two-coarse-grain-models">
<h2>Calculate the RMSD Between two Coarse-Grain Models<a class="headerlink" href="#calculate-the-rmsd-between-two-coarse-grain-models" title="Permalink to this headline">¶</a></h2>
<p>Like PDB chains, we can also compute an RMSD value for two coarse grain models.
For this we need to use the virtual residues of the helices as the atoms and
compute the RMSD value amongst them:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">forgi.threedee.model.coarse_grain</span> <span class="k">as</span> <span class="nn">ftmc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">forgi.threedee.model.similarity</span> <span class="k">as</span> <span class="nn">ftms</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cg1</span><span class="p">,</span> <span class="o">=</span> <span class="n">forgi</span><span class="o">.</span><span class="n">load_rna</span><span class="p">(</span><span class="s1">&#39;test/forgi/threedee/data/1y26.cg&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cg2</span><span class="p">,</span> <span class="o">=</span> <span class="n">forgi</span><span class="o">.</span><span class="n">load_rna</span><span class="p">(</span><span class="s1">&#39;test/forgi/threedee/data/1y26.cg&#39;</span><span class="p">)</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">ftms</span><span class="o">.</span><span class="n">cg_rmsd</span><span class="p">(</span><span class="n">cg1</span><span class="p">,</span><span class="n">cg2</span><span class="p">)</span>
<span class="go">0.0</span>
</pre></div>
</div>
<p>If we like more control, we can export any list of coordinates from the coarse grain moidel that
is in a defined order and calculate the rmsd between these two point clouds.
Examples would include the start and end coordinates of every stem for a faster RMSD estimation
that is less accurate but avoids the overhead of generating the virtual residues.</p>
<p>Note that currently forgi.threedee.model.similarity.rmsd centers and optimallly rotates the
two point clouds to minimize the RMSD. For an uncentered RMSD use
forgi.threedee.utilities.vector.vector_set_rmsd</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">forgi.threedee.model.coarse_grain</span> <span class="k">as</span> <span class="nn">ftmc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">forgi.threedee.utilities.graph_pdb</span> <span class="k">as</span> <span class="nn">ftug</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">forgi.threedee.model.similarity</span> <span class="k">as</span> <span class="nn">ftms</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cg1</span><span class="p">,</span> <span class="o">=</span> <span class="n">forgi</span><span class="o">.</span><span class="n">load_rna</span><span class="p">(</span><span class="s1">&#39;test/forgi/threedee/data/1y26.cg&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cg2</span><span class="p">,</span> <span class="o">=</span> <span class="n">forgi</span><span class="o">.</span><span class="n">load_rna</span><span class="p">(</span><span class="s1">&#39;test/forgi/threedee/data/1y26.cg&#39;</span><span class="p">)</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coords1</span> <span class="o">=</span> <span class="n">cg1</span><span class="o">.</span><span class="n">get_ordered_stem_poss</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coords2</span> <span class="o">=</span> <span class="n">cg2</span><span class="o">.</span><span class="n">get_ordered_stem_poss</span><span class="p">()</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="n">ftms</span><span class="o">.</span><span class="n">rmsd</span><span class="p">(</span><span class="n">coords1</span><span class="p">,</span> <span class="n">coords2</span><span class="p">)</span>
<span class="go">0.0</span>
</pre></div>
</div>
</div>
<div class="section" id="scalar-descriptors-of-rna-3d-structures">
<h2>Scalar descriptors of RNA 3D structures<a class="headerlink" href="#scalar-descriptors-of-rna-3d-structures" title="Permalink to this headline">¶</a></h2>
<p>To characterize an RNA 3D structures, a lot of descriptors could be potentially useful.
The <code class="code docutils literal"><span class="pre">forgi.threedee.model.descriptors</span></code> library includes the following scalar descriptors.</p>
<blockquote>
<div><ol class="arabic simple">
<li>The Radius of gyration</li>
<li>The anisotropy</li>
<li>The asphericity</li>
</ol>
</div></blockquote>
<p>Note that 1-3 are related, as they could all be derived from the gyration tensor.
An overview over these properties can be found for example in the introduction of
reference <a class="reference internal" href="#ref2"><span class="std std-ref">[2]</span></a> and in the papers referenced therein.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">forgi.threedee.model.coarse_grain</span> <span class="k">as</span> <span class="nn">ftmc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">forgi.threedee.model.descriptors</span> <span class="k">as</span> <span class="nn">ftmd</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cg</span><span class="p">,</span> <span class="o">=</span> <span class="n">forgi</span><span class="o">.</span><span class="n">load_rna</span><span class="p">(</span><span class="s1">&#39;test/forgi/threedee/data/1y26.cg&#39;</span><span class="p">)</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span> <span class="s2">&quot;Radius of Gyration&quot;</span><span class="p">,</span> <span class="n">cg</span><span class="o">.</span><span class="n">radius_of_gyration</span><span class="p">()</span> <span class="p">)</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coords</span> <span class="o">=</span> <span class="n">cg</span><span class="o">.</span><span class="n">get_ordered_stem_poss</span><span class="p">()</span> <span class="c1">#Use only coarse grained coordinates of stems</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coords</span> <span class="o">=</span> <span class="n">cg</span><span class="o">.</span><span class="n">get_ordered_virtual_residue_poss</span><span class="p">()</span> <span class="c1">#Alternatively use coordinates of virtual residues</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="s2">&quot;Anisotropy&quot;</span><span class="p">,</span> <span class="n">ftmd</span><span class="o">.</span><span class="n">anisotropy</span><span class="p">(</span><span class="n">coords</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span> <span class="s2">&quot;Asphericity&quot;</span><span class="p">,</span> <span class="n">ftmd</span><span class="o">.</span><span class="n">asphericity</span><span class="p">(</span><span class="n">coords</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="determine-if-two-atoms-are-covalently-bonded">
<h2>Determine if two Atoms are Covalently Bonded<a class="headerlink" href="#determine-if-two-atoms-are-covalently-bonded" title="Permalink to this headline">¶</a></h2>
<p>The difference between covalently bonded and unbonded atoms needs to be taken
into account when calculating clash scores. Covalently bonded atoms can be
closer to each other in a real structure than unbonded atoms. Based on the
identity of the atoms and their parent nucleotides, the <code class="docutils literal"><span class="pre">is_covalent</span></code>
function tries to determine the whether two atoms are covalently bonded
(not taking into account the geometry, but only the atom names). This only
works for unmodified nucleotides.</p>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">forgi.threedee.utilities.pdb</span> <span class="k">as</span> <span class="nn">ftup</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">ftup</span><span class="o">.</span><span class="n">get_biggest_chain</span><span class="p">(</span><span class="s1">&#39;test/forgi/threedee/data/2mis.pdb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ftup</span><span class="o">.</span><span class="n">is_covalent</span><span class="p">([</span><span class="n">c</span><span class="p">[</span><span class="mi">10</span><span class="p">][</span><span class="s2">&quot;C3&#39;&quot;</span><span class="p">],</span> <span class="n">c</span><span class="p">[</span><span class="mi">10</span><span class="p">][</span><span class="s2">&quot;C4&#39;&quot;</span><span class="p">]])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ftup</span><span class="o">.</span><span class="n">is_covalent</span><span class="p">([</span><span class="n">c</span><span class="p">[</span><span class="mi">10</span><span class="p">][</span><span class="s2">&quot;C3&#39;&quot;</span><span class="p">],</span> <span class="n">c</span><span class="p">[</span><span class="mi">10</span><span class="p">][</span><span class="s2">&quot;C5&#39;&quot;</span><span class="p">]])</span>
<span class="go">False</span>
</pre></div>
</div>
</div>
<div class="section" id="determine-the-total-rotation-of-a-stem">
<h2>Determine the total rotation of a stem<a class="headerlink" href="#determine-the-total-rotation-of-a-stem" title="Permalink to this headline">¶</a></h2>
<p>The location of a stem’s minor groove at the beginning and the end of the stem is stored as
the twist vector. To calculate the total rotation of a stem, we need to calculate the angle
between the two twist vectors and find out how many full turns around the helix we took.</p>
<p>An ideal helix turns 36.4 degrees per base-pair, so given the number of basepairs we can
calculate the number of full turns.</p>
</div>
<div class="section" id="citations">
<h2>Citations<a class="headerlink" href="#citations" title="Permalink to this headline">¶</a></h2>
<p id="ref1">[1] <em>Sandra Smit, Kristian Rother, Jaap Heringa, and Rob Knight</em>.
<strong>From knotted to nested RNA structures: a variety of computational methods for pseudoknot removal.</strong>
RNA (2008) 14(3):410-416.</p>
<p id="ref2">[2] <em>Handan Arkın and Wolfhard Janke</em>.
<strong>Gyration tensor based analysis of the shapes of polymer chains in an attractive spherical cage</strong>
The Journal of Chemical Physics (2013) 138:054904, <a class="reference external" href="http://dx.doi.org/10.1063/1.4788616">http://dx.doi.org/10.1063/1.4788616</a></p>
</div>
</div>


      </div>
      <div class="bottomnav">
      
        <p>
        «&#160;&#160;<a href="graph_tutorial.html">RNA Secondary Structure as a Graph Using the forgi Library</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="apidoc/forgi.html">Forgi API documentation</a>&#160;&#160;»
        </p>

      </div>

    <div class="footer" role="contentinfo">
        &#169; Copyright 2012-2016 Peter Kerpedjiev &lt;pkerp@tbi.univie.ac.at&gt;; 2015-2018 Bernhard Thiel &lt;thiel@tbi.univie.ac.at&gt;.
      Last updated on Feb 25, 2019.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.7.
    </div>
  </body>
</html>